gtn/functions/compose.cpp

```diff
@@ -450,3 +456,931 @@ Graph compose(
+    // Note: Multiple threads can have the same dstIdx and write to the same
+    //       location and collide. This _should_ be fine since they are going
+    //       to write the same value
+    newGraphDP.start[newNodesOffset[dstIdx]] = dstNodeStartAndAccept.first;
+    newGraphDP.accept[newNodesOffset[dstIdx]] = dstNodeStartAndAccept.second;
+
+    // Both of these increments are atomic
+    int inArcIdx = newGraphDP.inArcOffset[newNodesOffset[dstIdx]]++;
+    int outArcIdx = newGraphDP.outArcOffset[newNodesOffset[curIdx]]++;
+
+    // outArcIdx is also the arc identifier
+    newGraphDP.outArcs[outArcIdx] = outArcIdx;
+    newGraphDP.inArcs[inArcIdx] = outArcIdx;
+
+    // Fill in everything else for this arc
+    newGraphDP.ilabels[outArcIdx] = ilabel;
+    newGraphDP.olabels[outArcIdx] = olabel;
+    newGraphDP.srcNodes[outArcIdx] = newNodesOffset[curIdx];
+    newGraphDP.dstNodes[outArcIdx] = newNodesOffset[dstIdx];
+    newGraphDP.weights[outArcIdx] = weight;
+
+    gradInfo.first[outArcIdx] = arcPair.first;
+    gradInfo.second[outArcIdx] = arcPair.second;
+  }
+}
+
+// Convert bool array two pairs for true flags
+std::vector<std::pair<int, int>> convertToNodePair(
+    const std::vector<bool>& flags,
+    int extent) {
+  std::vector<std::pair<int, int>> toExploreNodePair;
+  for (size_t i = 0; i < flags.size(); ++i) {
+    if (flags[i] == true) {
+      toExploreNodePair.push_back(OneDToTwoDIndex(i, extent));
+    }
+  }
+
+  return toExploreNodePair;
+}
+
+// Takes a bool array with flags set for nodes to pick and returns
+// an array with indices that were set as true
+std::vector<int> convertToNodes(const std::vector<bool>& flags) {
+  std::vector<int> nodes;
+
+  for (size_t i = 0; i < flags.size(); ++i) {
+    if (flags[i]) {
+      nodes.push_back(i);
+    }
+  }
+
+  return nodes;
+}
+
+std::tuple<std::pair<bool, bool>, std::pair<bool, bool>> getStartAndAccept(
+    const GraphDataParallel& graphDP1,
+    const GraphDataParallel& graphDP2,
+    const std::pair<int, int>& srcNodePair,
+    const std::pair<int, int>& dstNodePair) {
+  const std::pair<bool, bool> srcNodeStartAndAccept = std::make_pair(
+      graphDP1.start[srcNodePair.first] && graphDP2.start[srcNodePair.second],
+      graphDP1.accept[srcNodePair.first] &&
+          graphDP2.accept[srcNodePair.second]);
+
+  const std::pair<bool, bool> dstNodeStartAndAccept = std::make_pair(
+      graphDP1.start[dstNodePair.first] && graphDP2.start[dstNodePair.second],
+      graphDP1.accept[dstNodePair.first] &&
+          graphDP2.accept[dstNodePair.second]);
+
+  return std::make_tuple(srcNodeStartAndAccept, dstNodeStartAndAccept);
+}
+
+} // namespace
+
+// Convert from AOS to SOA
+// Assumption: nodes are numbered [0..graph.numNodes())
+//           : arcs are numbered [0..graph.numArcs())
+GraphDataParallel convertToDataParallel(const Graph& graph) {
+  GraphDataParallel graphDP;
+
+  graphDP.accept.resize(graph.numNodes());
```

Comment:
`graphDP.accept.resize(graph.numNodes(), false);` can avoid the extra call `std::fill`.

Reply:
Thanks!! will add to `resize` after adding `assert`